site-src/geps/gep-724.md

```diff
@@ -0,0 +1,494 @@
+# GEP-724: Refresh Route-Gateway Binding
+
+* Issue URL: [#724](https://github.com/kubernetes-sigs/gateway-api/issues/724)
+* Status: Implementable
+
+## TLDR
+
+This GEP proposes changes to Route-Gateway binding that will result in Routes
+attaching to Gateways with direct references. When supporting Routes in multiple
+namespaces, Gateways will need to specify the namespaces they trust Routes in.
+These changes will slightly simplify the Route-Gateway relationship and make way
+for the future addition of Route inclusion (Routes including other Routes).
+
+## Goals
+
+Refactor cross-namespace Route-Gateway binding to:
+
+* Be more consistent with [cross-namespace references from
+  Routes](/geps/gep-709.md)
+* Provide a clear path to enable Route inclusion (Routes including Routes).
+* Simplify user experience based on initial feedback.
+* Enable other kinds of Route parents in addition to Gateway, this could include:
+    * Routes (as part of Route inclusion)
```

Comment:
Agree with the risks of implementing Route inclusion this way. One distinction I should probably make more clear is that I see this as a potential advantage of the flexibility of this approach, not that this should ultimately be how we implement Route inclusion. Maybe we do need to start on a GEP for Route inclusion so we can get the final piece of this puzzle solved, even if it doesn't make it into the initial v1alpha2 release...

I think it's finding itself in an awkward place between ReferencePolicy and Route->Gateway, both patterns are very similar and could potentially apply to Route->Route, but based on what you're saying here and the comment from @howardjohn above, ReferencePolicy may make more sense for that relationship. If that's the case, we just need to be sure that this `attachTo` concept is more clearly scoped to either just Gateway, or Gateway/Mesh-like parent resources.

Reply:
I think the 1:many scenario described in a different [comment](https://github.com/kubernetes-sigs/gateway-api/pull/725#discussion_r672455899) is really the only way parentRef could make sense to refer/attach to a parent Route. With that kind of approach, we'd mirror the Route-Gateway relationship pretty closely for Route-Route. That has some appeal in that it would allow for broader delegation of a path to a namespace and not just a specific route.

On the other hand, it would add a lot of complexity to an already complex API. The other alternative of direct refs to Routes to include with ReferencePolicy as the trust mechanism feels _much_ simpler. Maybe that means we need some kind of more specific word here to indicate "resource that exists conceptually above Routing". For 99% of use cases, I think that's just Gateway, but I do like the idea of a term that's generic enough to represent a reference to a mesh or other similar non-Gateway concept.

Reply:
One positive here is that renaming a field is technically backwards compatible, so we do have some flexibility here if we don't get the name right the first time.

Reply:
The thing that makes Route -> Route inclusion much more complex is the semantics about how you carve up the URL/request metadata space. In HTTPProxy, we explicitly decided that the rules for assigning certain requests to a child HTTPProxy (in the Gateway API, this would be a child HTTPRoute) would become surrounding, uneditable *constraints* on that child, and that the child *did not need to know* about those constraints beforehand.

This has tradeoffs, obviously, but the key part here is that you can't design the Route -> Route reference mechanism without also designing how much the included Routes will know about where they are included.

In short, I think that what we have here is broad enough to *allow* us to experiment in either way, and we shouldn't hold this GEP up on figuring out how Route->Route inclusion works.

Reply:
Thanks for the great summary @youngnick! I think we're all leaning in the same direction of not trying to be too prescriptive about how Route inclusion could work, but instead leave this as one potential approach.  I've updated this line in the GEP to read "as part of one potential approach to Route inclusion".